components/scream/src/share/io/tests/io.cpp

```diff
@@ -251,116 +264,265 @@ TEST_CASE("input_output_basic","io")
-  f2.sync_to_host();
-  f3.sync_to_host();
-  for (int ii=0;ii<num_lcols;++ii) {
-    REQUIRE(std::abs(f1_host(ii)-(max_steps*dt+ii))<tol);
+  // We can use the produced output files to simultaneously check output quality and the
+  // ability to read input.
+  // NOTE: single-snap file start outputing at the final time only,
+  //       while multi-snap starts to output at the first time step
+  // First we re-construct the timestamp for the input file:
+  REQUIRE(output_stamps.size()>0);
+  auto input_params = get_in_params(output_type,output_freq_units, io_comm,output_stamps.front());
+  const Real tol = 1000*std::numeric_limits<Real>::epsilon();
+
+  // TODO: Create a small nc dummy file and a separate unit test which tests all input functions.
+  // Test that pio_inq_dimlen is correct, using a file from one of the above parameter lists.
+  {
+    auto test_filename = input_params.get<std::string>("Filename");
+    scorpio::register_file(test_filename,scorpio::Read);
+    Int test_gcols_len = scorpio::get_dimlen_c2f(test_filename.c_str(),"ncol");
+    REQUIRE(test_gcols_len==num_gcols);
+    scorpio::eam_pio_closefile(test_filename);
+  }
+
+  auto f1 = field_manager->get_field("field_1");
+  auto f2 = field_manager->get_field("field_2");
+  auto f3 = field_manager->get_field("field_3");
+  auto f4 = field_manager->get_field("field_packed");
+  auto f1_host = f1.get_view<Real*,Host>();
+  auto f2_host = f2.get_view<Real*,Host>();
+  auto f3_host = f3.get_view<Real**,Host>();
+  auto f4_host = f4.get_view<Real**,Host>();
+
+
+  if (output_type == "instant") {
+    // Check instant output
+    AtmosphereInput ins_input(input_params,field_manager);
+    ins_input.read_variables();
+    f1.sync_to_host();
+    f2.sync_to_host();
+    f3.sync_to_host();
+    f4.sync_to_host();
+  
+    // The diagnostic is not present in the field manager.  So we can't use the scorpio_input class
+    // to read in the data.  Here we use raw IO routines to gather the data for testing.
+    auto f_diag_ins_h = get_diagnostic_input(io_comm, gm, 1, input_params.get<std::string>("Filename"));
+  
+    for (int ii=0;ii<num_lcols;++ii) {
+      REQUIRE(std::abs(f1_host(ii)-(max_steps*dt+ii))<tol);
+      for (int jj=0;jj<num_levs;++jj) {
+        REQUIRE(std::abs(f3_host(ii,jj)-(ii+max_steps*dt + (jj+1)/10.))<tol);
+        REQUIRE(std::abs(f4_host(ii,jj)-(ii+max_steps*dt + (jj+1)/10.))<tol);
+        REQUIRE(std::abs(f_diag_ins_h(ii,jj)-(3.0*f4_host(ii,jj)+2.0))<tol);
+      }
+    }
     for (int jj=0;jj<num_levs;++jj) {
-      REQUIRE(std::abs(f3_host(ii,jj)-(ii+max_steps*dt + (jj+1)/10.))<tol);
+      REQUIRE(std::abs(f2_host(jj)-(max_steps*dt + (jj+1)/10.))<tol);
+    }
+    ins_input.finalize();
+    reset_fields();
+  } else if (output_type == "average") {
+    // Check average output
+    AtmosphereInput avg_input(input_params,field_manager);
+    avg_input.read_variables();
+    f1.sync_to_host();
+    f2.sync_to_host();
+    f3.sync_to_host();
```

Comment:
Any reason why f4 and the diag output are not checked for output_type other than "instant"?

Reply:
ahhh good catch.  I think that f4 was just to add a test for a packed variable so I only originally had it testing in one case.  But it was trivial to have it tested in all cases.